### ISSUE TITLE
High CPU Usage when in watch (Watched.clearWhenTriggered every 0.1s)

### ISSUE THREAD

**Issue by matthughes** [sbt/sbt#1862]
After running SBT "~testQuick", my CPU started to take off.  It's happened to me a couple times and seems to be correlated with how long I've had the console open.  This isn't a large project (~50 source file).

Sampling the process through VisualVM, I noticed the following:

SBT (by way of Mina) was allocating a huge number of throwaway HashMaps$KeyIterator.  In my system, it was allocating 156 million `HashMap$KeyIterator` objects every 2-3 seconds; then GC would kick in and collect them all.  I have 1024mb allocated to the heap.  The GC kicks in around 400mb so I don't think adding more memory would help.  The CPU was pegged (all 8 cores) in the following methods:

```
NioProcessor.select() 60% 
NioProcessor.wakeup() 40%
```

I was never able to capture the stack trace where it was allocating, but this issue looks very similar: http://mail-archives.apache.org/mod_mbox/mina-users/200808.mbox/%3CF6E283B31FEA204AB1DE54059FA622E306A02B29@ecnshmw511.eapac.ericsson.se%3E

What was really strange was that even after I exited out of the watch -- just sitting in console doing nothing -- the polling continued causing the performance problems.   

Running:

```
SBT 0.13.7
OS X 10.10.2
java version "1.8.0_20"
Java(TM) SE Runtime Environment (build 1.8.0_20-b26)
```

Only interesting thing in my SBT profile might be:

```
triggeredMessage := Watched.clearWhenTriggered
// Reduce from 500 default to 100
pollInterval := 100
```

**Comment by matthughes** [comment]
Here is a thread dump _after_ exiting the watch.  You can see there are still 6 NIO threads polling.  Only way to get CPU under control at this point is to completely restart SBT.

```
› jstack 50586
2015-02-17 10:41:41
Full thread dump Java HotSpot(TM) 64-Bit Server VM (25.20-b23 mixed mode):

"process reaper" #30873 daemon prio=10 os_prio=31 tid=0x00007fe307850000 nid=0xb7f3 waiting on condition [0x000000011076e000]
   java.lang.Thread.State: TIMED_WAITING (parking)
  at sun.misc.Unsafe.park(Native Method)
  - parking to wait for  <0x0000000780002b90> (a java.util.concurrent.SynchronousQueue$TransferStack)
  at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
  at java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:460)
  at java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:362)
  at java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:941)
  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1066)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)

"Attach Listener" #30733 daemon prio=9 os_prio=31 tid=0x00007fe303f30800 nid=0x986b waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"NioProcessor-6" #23289 prio=5 os_prio=31 tid=0x00007fe309740800 nid=0xc70b runnable [0x00000001227c5000]
   java.lang.Thread.State: RUNNABLE
  at sun.nio.ch.KQueueArrayWrapper.interrupt(Native Method)
  at sun.nio.ch.KQueueArrayWrapper.interrupt(KQueueArrayWrapper.java:203)
  at sun.nio.ch.KQueueSelectorImpl.wakeup(KQueueSelectorImpl.java:243)
  - locked <0x000000078a7b4b50> (a java.lang.Object)
  at org.apache.mina.transport.socket.nio.NioProcessor.wakeup(NioProcessor.java:113)
  at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.run(AbstractPollingIoProcessor.java:1163)
  at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:64)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)

"NioProcessor-5" #23283 prio=5 os_prio=31 tid=0x00007fe30cbca000 nid=0xab1f runnable [0x0000000120227000]
   java.lang.Thread.State: RUNNABLE
  at sun.nio.ch.KQueueArrayWrapper.kevent0(Native Method)
  at sun.nio.ch.KQueueArrayWrapper.poll(KQueueArrayWrapper.java:198)
  at sun.nio.ch.KQueueSelectorImpl.doSelect(KQueueSelectorImpl.java:103)
  at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
  - locked <0x000000078a802ca0> (a sun.nio.ch.Util$2)
  - locked <0x000000078a802c90> (a java.util.Collections$UnmodifiableSet)
  - locked <0x000000078a802950> (a sun.nio.ch.KQueueSelectorImpl)
  at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
  at org.apache.mina.transport.socket.nio.NioProcessor.select(NioProcessor.java:97)
  at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.run(AbstractPollingIoProcessor.java:1074)
  at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:64)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)

"NioProcessor-3" #22649 prio=5 os_prio=31 tid=0x00007fe30d2d8800 nid=0x9a9f runnable [0x000000011ff24000]
   java.lang.Thread.State: RUNNABLE
  at sun.nio.ch.KQueueArrayWrapper.kevent0(Native Method)
  at sun.nio.ch.KQueueArrayWrapper.poll(KQueueArrayWrapper.java:198)
  at sun.nio.ch.KQueueSelectorImpl.doSelect(KQueueSelectorImpl.java:103)
  at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
  - locked <0x0000000789e68ec0> (a sun.nio.ch.Util$2)
  - locked <0x0000000789e68ed0> (a java.util.Collections$UnmodifiableSet)
  - locked <0x0000000789e68e70> (a sun.nio.ch.KQueueSelectorImpl)
  at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
  at org.apache.mina.transport.socket.nio.NioProcessor.select(NioProcessor.java:97)
  at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.run(AbstractPollingIoProcessor.java:1074)
  at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:64)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)

"NioProcessor-1" #22608 prio=5 os_prio=31 tid=0x00007fe305642800 nid=0x720b runnable [0x000000011fb18000]
   java.lang.Thread.State: RUNNABLE
  at sun.nio.ch.KQueueArrayWrapper.kevent0(Native Method)
  at sun.nio.ch.KQueueArrayWrapper.poll(KQueueArrayWrapper.java:198)
  at sun.nio.ch.KQueueSelectorImpl.doSelect(KQueueSelectorImpl.java:103)
  at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
  - locked <0x0000000789e9d368> (a sun.nio.ch.Util$2)
  - locked <0x0000000789e9d378> (a java.util.Collections$UnmodifiableSet)
  - locked <0x0000000789e9d318> (a sun.nio.ch.KQueueSelectorImpl)
  at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
  at org.apache.mina.transport.socket.nio.NioProcessor.select(NioProcessor.java:97)
  at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.run(AbstractPollingIoProcessor.java:1074)
  at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:64)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)

"NioProcessor-2" #22601 prio=5 os_prio=31 tid=0x00007fe304a35000 nid=0x8c0f runnable [0x000000011f403000]
   java.lang.Thread.State: RUNNABLE
  at sun.nio.ch.KQueueArrayWrapper.kevent0(Native Method)
  at sun.nio.ch.KQueueArrayWrapper.poll(KQueueArrayWrapper.java:198)
  at sun.nio.ch.KQueueSelectorImpl.doSelect(KQueueSelectorImpl.java:103)
  at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
  - locked <0x0000000789e88878> (a sun.nio.ch.Util$2)
  - locked <0x0000000789e88868> (a java.util.Collections$UnmodifiableSet)
  - locked <0x0000000789e88628> (a sun.nio.ch.KQueueSelectorImpl)
  at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
  at org.apache.mina.transport.socket.nio.NioProcessor.select(NioProcessor.java:97)
  at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.run(AbstractPollingIoProcessor.java:1074)
  at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:64)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)

"NioSocketAcceptor-1" #22597 prio=5 os_prio=31 tid=0x00007fe30450a000 nid=0x960b in Object.wait() [0x000000011c5d5000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
  at java.lang.Object.wait(Native Method)
  at org.apache.mina.core.future.DefaultIoFuture.await0(DefaultIoFuture.java:192)
  - locked <0x0000000789e95668> (a org.apache.mina.core.future.DefaultIoFuture)
  at org.apache.mina.core.future.DefaultIoFuture.awaitUninterruptibly(DefaultIoFuture.java:133)
  at org.apache.mina.core.polling.AbstractPollingIoProcessor.dispose(AbstractPollingIoProcessor.java:192)
  at org.apache.mina.core.service.SimpleIoProcessorPool.dispose(SimpleIoProcessorPool.java:327)
  - locked <0x0000000789e956e8> (a java.lang.Object)
  at org.apache.mina.core.polling.AbstractPollingIoAcceptor$Acceptor.run(AbstractPollingIoAcceptor.java:506)
  at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:64)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)

"NioProcessor-4" #3114 prio=5 os_prio=31 tid=0x00007fe306401000 nid=0x992f runnable [0x000000011d61c000]
   java.lang.Thread.State: RUNNABLE
  at sun.nio.ch.IOUtil.drain(Native Method)
  at sun.nio.ch.KQueueSelectorImpl.updateSelectedKeys(KQueueSelectorImpl.java:167)
  - locked <0x0000000782f4fdc0> (a java.lang.Object)
  at sun.nio.ch.KQueueSelectorImpl.doSelect(KQueueSelectorImpl.java:108)
  at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
  - locked <0x0000000782f4fcc8> (a sun.nio.ch.Util$2)
  - locked <0x0000000782f4fcb8> (a java.util.Collections$UnmodifiableSet)
  - locked <0x0000000782f4fa68> (a sun.nio.ch.KQueueSelectorImpl)
  at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
  at org.apache.mina.transport.socket.nio.NioProcessor.select(NioProcessor.java:97)
  at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.run(AbstractPollingIoProcessor.java:1074)
  at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:64)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)

"Service Thread" #9 daemon prio=9 os_prio=31 tid=0x00007fe30380b800 nid=0x5103 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C1 CompilerThread3" #8 daemon prio=9 os_prio=31 tid=0x00007fe30483e800 nid=0x4f03 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread2" #7 daemon prio=9 os_prio=31 tid=0x00007fe30380d800 nid=0x4d03 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread1" #6 daemon prio=9 os_prio=31 tid=0x00007fe30483e000 nid=0x4b03 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #5 daemon prio=9 os_prio=31 tid=0x00007fe304008000 nid=0x4903 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Signal Dispatcher" #4 daemon prio=9 os_prio=31 tid=0x00007fe304005800 nid=0x3913 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" #3 daemon prio=8 os_prio=31 tid=0x00007fe30485e800 nid=0x3503 in Object.wait() [0x0000000110693000]
   java.lang.Thread.State: WAITING (on object monitor)
  at java.lang.Object.wait(Native Method)
  at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:142)
  - locked <0x00000007800140a0> (a java.lang.ref.ReferenceQueue$Lock)
  at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:158)
  at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)

"Reference Handler" #2 daemon prio=10 os_prio=31 tid=0x00007fe305005000 nid=0x3303 in Object.wait() [0x0000000110590000]
   java.lang.Thread.State: WAITING (on object monitor)
  at java.lang.Object.wait(Native Method)
  at java.lang.Object.wait(Object.java:502)
  at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:157)
  - locked <0x000000078000eba0> (a java.lang.ref.Reference$Lock)

"main" #1 prio=5 os_prio=31 tid=0x00007fe304801000 nid=0x1303 runnable [0x0000000100d5e000]
   java.lang.Thread.State: RUNNABLE
  at java.io.FileInputStream.read0(Native Method)
  at java.io.FileInputStream.read(FileInputStream.java:198)
  at jline.internal.NonBlockingInputStream.read(NonBlockingInputStream.java:248)
  at jline.internal.InputStreamReader.read(InputStreamReader.java:261)
  - locked <0x00000007bff20200> (a jline.internal.NonBlockingInputStream)
  at jline.internal.InputStreamReader.read(InputStreamReader.java:198)
  - locked <0x00000007bff20200> (a jline.internal.NonBlockingInputStream)
  at jline.console.ConsoleReader.readCharacter(ConsoleReader.java:2038)
  at jline.console.ConsoleReader.readLine(ConsoleReader.java:2242)
  at jline.console.ConsoleReader.readLine(ConsoleReader.java:2162)
  at sbt.JLine.sbt$JLine$$readLineDirectRaw(LineReader.scala:42)
  at sbt.JLine$$anonfun$readLineDirect$2.apply(LineReader.scala:34)
  at sbt.JLine$$anonfun$readLineDirect$2.apply(LineReader.scala:34)
  at sbt.Signals0.withHandler(Signal.scala:81)
  at sbt.Signals$.withHandler(Signal.scala:11)
  at sbt.JLine.readLineDirect(LineReader.scala:34)
  at sbt.JLine.readLineWithHistory(LineReader.scala:27)
  at sbt.JLine.sbt$JLine$$unsynchronizedReadLine(LineReader.scala:19)
  at sbt.JLine$$anonfun$readLine$1.apply(LineReader.scala:16)
  at sbt.JLine$$anonfun$readLine$1.apply(LineReader.scala:16)
  at sbt.JLine$$anonfun$withJLine$1.apply(LineReader.scala:114)
  at sbt.JLine$$anonfun$withJLine$1.apply(LineReader.scala:112)
  at sbt.JLine$.withTerminal(LineReader.scala:86)
  - locked <0x00000007bff16020> (a jline.UnixTerminal)
  - locked <0x00000007806acc88> (a sbt.JLine$)
  at sbt.JLine$.withJLine(LineReader.scala:112)
  at sbt.JLine.readLine(LineReader.scala:16)
  at sbt.BasicCommands$$anonfun$shell$1.apply(BasicCommands.scala:164)
  at sbt.BasicCommands$$anonfun$shell$1.apply(BasicCommands.scala:160)
  at sbt.Command$$anonfun$command$1$$anonfun$apply$1.apply(Command.scala:29)
  at sbt.Command$$anonfun$command$1$$anonfun$apply$1.apply(Command.scala:29)
  at sbt.Command$.process(Command.scala:92)
  at sbt.MainLoop$$anonfun$1$$anonfun$apply$1.apply(MainLoop.scala:98)
  at sbt.MainLoop$$anonfun$1$$anonfun$apply$1.apply(MainLoop.scala:98)
  at sbt.State$$anon$1.process(State.scala:184)
  at sbt.MainLoop$$anonfun$1.apply(MainLoop.scala:98)
  at sbt.MainLoop$$anonfun$1.apply(MainLoop.scala:98)
  at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
  at sbt.MainLoop$.next(MainLoop.scala:98)
  at sbt.MainLoop$.run(MainLoop.scala:91)
  at sbt.MainLoop$$anonfun$runWithNewLog$1.apply(MainLoop.scala:70)
  at sbt.MainLoop$$anonfun$runWithNewLog$1.apply(MainLoop.scala:65)
  at sbt.Using.apply(Using.scala:24)
  at sbt.MainLoop$.runWithNewLog(MainLoop.scala:65)
  at sbt.MainLoop$.runAndClearLast(MainLoop.scala:48)
  at sbt.MainLoop$.runLoggedLoop(MainLoop.scala:32)
  at sbt.MainLoop$.runLogged(MainLoop.scala:24)
  at sbt.StandardMain$.runManaged(Main.scala:53)
  at sbt.xMain.run(Main.scala:28)
  at xsbt.boot.Launch$$anonfun$run$1.apply(Launch.scala:109)
  at xsbt.boot.Launch$.withContextLoader(Launch.scala:129)
  at xsbt.boot.Launch$.run(Launch.scala:109)
  at xsbt.boot.Launch$$anonfun$apply$1.apply(Launch.scala:36)
  at xsbt.boot.Launch$.launch(Launch.scala:117)
  at xsbt.boot.Launch$.apply(Launch.scala:19)
  at xsbt.boot.Boot$.runImpl(Boot.scala:44)
  at xsbt.boot.Boot$.main(Boot.scala:20)
  at xsbt.boot.Boot.main(Boot.scala)

"VM Thread" os_prio=31 tid=0x00007fe30485e000 nid=0x3103 runnable

"GC task thread#0 (ParallelGC)" os_prio=31 tid=0x00007fe30480e000 nid=0x2103 runnable

"GC task thread#1 (ParallelGC)" os_prio=31 tid=0x00007fe305000000 nid=0x2303 runnable

"GC task thread#2 (ParallelGC)" os_prio=31 tid=0x00007fe30480f000 nid=0x2503 runnable

"GC task thread#3 (ParallelGC)" os_prio=31 tid=0x00007fe30480f800 nid=0x2703 runnable

"GC task thread#4 (ParallelGC)" os_prio=31 tid=0x00007fe304810000 nid=0x2903 runnable

"GC task thread#5 (ParallelGC)" os_prio=31 tid=0x00007fe304810800 nid=0x2b03 runnable

"GC task thread#6 (ParallelGC)" os_prio=31 tid=0x00007fe304811800 nid=0x2d03 runnable

"GC task thread#7 (ParallelGC)" os_prio=31 tid=0x00007fe304812000 nid=0x2f03 runnable

"VM Periodic Task Thread" os_prio=31 tid=0x00007fe30380c800 nid=0x5303 waiting on condition

JNI global references: 313
```

**Comment by eed3si9n** [comment]
So you are

a.) making jline clear the screen
b.) every 0.1 second
c.) on Java 8

and things go haywire? Can you tweak the parameters around, like not clearing the screen, every 0.5 like the default, or using Java 7. Also were you able to figure out where all the `HashMap$KeyIterator` is coming from?

**Comment by eed3si9n** [comment]
Triggered execution on macOS will be fixed in 1.1.3. sbt/io#138
